From e4690f423f1aa0c4691dbd097f946b35ae96eeb2 Mon Sep 17 00:00:00 2001
From: Helge Deller <deller@gmx.de>
Date: Sat, 26 Nov 2022 21:29:31 +0100
Subject: [PATCH] parisc: Fix locking in pdc_iodc_print() firmware call

Utilize pdc_lock spinlock to protect parallel modifications of the
iodc_dbuf[] buffer, check length to prevent buffer overflow of
iodc_dbuf[], drop the iodc_retbuf[] buffer and fix some wrong
indentings.

Signed-off-by: Helge Deller <deller@gmx.de>
Cc: <stable@vger.kernel.org> # 6.0+
---
 arch/parisc/kernel/firmware.c | 70 ++++++++++++++++++-----------------
 1 file changed, 36 insertions(+), 34 deletions(-)

diff --git a/arch/parisc/kernel/firmware.c b/arch/parisc/kernel/firmware.c
index 1d69e8cf459c..8903455354c6 100644
--- a/arch/parisc/kernel/firmware.c
+++ b/arch/parisc/kernel/firmware.c
@@ -1233,9 +1233,8 @@ void pdc_io_reset_devices(void)
 
 #endif /* defined(BOOTLOADER) */
 
-/* locked by pdc_console_lock */
-static int __attribute__((aligned(8)))   iodc_retbuf[32];
-static char __attribute__((aligned(64))) iodc_dbuf[4096];
+/* locked by pdc_lock */
+static char iodc_dbuf[4096] __page_aligned_bss;
 
 /**
  * pdc_iodc_print - Console print using IODC.
@@ -1249,33 +1248,35 @@ static char __attribute__((aligned(64))) iodc_dbuf[4096];
  */
 int pdc_iodc_print(const unsigned char *str, unsigned count)
 {
-	unsigned int i, found = 0;
-	unsigned long flags;
+        unsigned int i, found = 0;
+        unsigned long flags;
 
-	for (i = 0; i < count;) {
-		switch(str[i]) {
-		case '\n':
-			iodc_dbuf[i+0] = '\r';
-			iodc_dbuf[i+1] = '\n';
-			i += 2;
-			found = 1;
-			goto print;
-		default:
-			iodc_dbuf[i] = str[i];
-			i++;
-			break;
-		}
-	}
+        count = min_t(unsigned int, count, sizeof(iodc_dbuf));
 
-print:
         spin_lock_irqsave(&pdc_lock, flags);
+        for (i = 0; i < count;) {
+                switch(str[i]) {
+                case '\n':
+                        iodc_dbuf[i+0] = '\r';
+                        iodc_dbuf[i+1] = '\n';
+                        i += 2;
+                        found = 1;
+                        goto print;
+                default:
+                        iodc_dbuf[i] = str[i];
+                        i++;
+                        break;
+                }
+        }
+
+print:
         real32_call(PAGE0->mem_cons.iodc_io,
                     (unsigned long)PAGE0->mem_cons.hpa, ENTRY_IO_COUT,
                     PAGE0->mem_cons.spa, __pa(PAGE0->mem_cons.dp.layers),
-                    __pa(iodc_retbuf), 0, __pa(iodc_dbuf), i, 0);
+                    __pa(pdc_result), 0, __pa(iodc_dbuf), i, 0);
         spin_unlock_irqrestore(&pdc_lock, flags);
 
-	return i - found;
+        return i - found;
 }
 
 #if !defined(BOOTLOADER)
@@ -1296,20 +1297,21 @@ int pdc_iodc_getc(void)
 		return 0;
 	
 	/* wait for a keyboard (rs232)-input */
-	spin_lock_irqsave(&pdc_lock, flags);
-	real32_call(PAGE0->mem_kbd.iodc_io,
-		    (unsigned long)PAGE0->mem_kbd.hpa, ENTRY_IO_CIN,
-		    PAGE0->mem_kbd.spa, __pa(PAGE0->mem_kbd.dp.layers), 
-		    __pa(iodc_retbuf), 0, __pa(iodc_dbuf), 1, 0);
+        spin_lock_irqsave(&pdc_lock, flags);
+        real32_call(PAGE0->mem_kbd.iodc_io,
+                    (unsigned long)PAGE0->mem_kbd.hpa, ENTRY_IO_CIN,
+                    PAGE0->mem_kbd.spa, __pa(PAGE0->mem_kbd.dp.layers),
+                    __pa(pdc_result), 0, __pa(iodc_dbuf), 1, 0);
+
+        ch = *iodc_dbuf;
+        /* like convert_to_wide() but for first return value only: */
+        status = *(int *)&pdc_result;
+        spin_unlock_irqrestore(&pdc_lock, flags);
 
-	ch = *iodc_dbuf;
-	status = *iodc_retbuf;
-	spin_unlock_irqrestore(&pdc_lock, flags);
+        if (status == 0)
+            return -1;
 
-	if (status == 0)
-	    return -1;
-	
-	return ch;
+        return ch;
 }
 
 int pdc_sti_call(unsigned long func, unsigned long flags,
-- 
2.43.0

